reframe/frontend/cli.py

```diff
@@ -920,9 +920,19 @@ def print_infoline(param, value):
                 testcases = filter(filters.have_not_name(name), testcases)
 
         if options.names:
-            testcases = filter(
-                filters.have_name('|'.join(options.names)), testcases
-            )
+            def _filter_multiple_names(names):
+                def _fn(case):
+                    return any(filters.have_name(n)(case) for n in names)
+
+                return _fn
+
+            # Differentiate between exact names and regexes since
+            # exact names cannot be combined in a single regex
+            exact_names = [n for n in options.names if '@' in n]
+            regex_names = [n for n in options.names if '@' not in n]
+
+            names = exact_names + ['|'.join(regex_names)]
+            testcases = filter(_filter_multiple_names(names), testcases)
```

Comment:
I think that we should rather introduce a separate filter just for matching the unique names. So you should split the names into `exact` and `regex` as you do now (in a single loop preferably), join the `regex_names` with `|` and use the current `have_name()` (but remove from it the logic about matching exact names) and then call another filter for matching any of the exact names.